src/System.Diagnostics.StackTrace/src/System/Diagnostics/StackTraceSymbols.CoreCLR.cs

```diff
@@ -2,33 +2,33 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using System.Collections.Concurrent;
 using System.IO;
 using System.Reflection;
 using System.Reflection.Metadata;
 using System.Reflection.Metadata.Ecma335;
 using System.Reflection.PortableExecutable;
+using System.Runtime.CompilerServices;
 
 namespace System.Diagnostics
 {
     internal class StackTraceSymbols : IDisposable
     {
-        private readonly ConcurrentDictionary<IntPtr, MetadataReaderProvider> _metadataCache;
+        private readonly ConditionalWeakTable<Assembly, MetadataReaderProvider> _metadataCache;
 
         /// <summary>
         /// Create an instance of this class.
         /// </summary>
         public StackTraceSymbols()
         {
-            _metadataCache = new ConcurrentDictionary<IntPtr, MetadataReaderProvider>();
+            _metadataCache = new ConditionalWeakTable<Assembly, MetadataReaderProvider>();
         }
 
         /// <summary>
         /// Clean up any cached providers.
         /// </summary>
         void IDisposable.Dispose()
         {
-            foreach (MetadataReaderProvider provider in _metadataCache.Values)
+            foreach ((Assembly cacheKey, MetadataReaderProvider provider) in _metadataCache)
```

Comment:
I wonder if the wildcard `_` could be used here instead of the `cacheKey` that we don't use.